http-clients/aws-crt-client/src/main/java/software/amazon/awssdk/http/crt/AwsCrtAsyncHttpClient.java

```diff
@@ -0,0 +1,383 @@
+                connPool = newConnPool;
+            } else {
+                // Multiple threads trying to open connections to the same URI at once, close the newer one
+                newConnPool.close();
+                connPool = alreadyExistingConnPool;
+            }
+        }
+
+        return connPool;
+    }
+
+    private List<HttpHeader> createHttpHeaderList(URI uri, AsyncExecuteRequest asyncRequest) {
+        SdkHttpRequest sdkRequest = asyncRequest.request();
+        List<HttpHeader> crtHeaderList = new ArrayList<>(sdkRequest.headers().size() + 2);
+
+        // Set Host Header if needed
+        if (isNullOrEmpty(sdkRequest.headers().get(HOST_HEADER))) {
+            crtHeaderList.add(new HttpHeader(HOST_HEADER, uri.getHost()));
+        }
+
+        // Add Connection Keep Alive Header to reuse this Http Connection as long as possible
+        if (isNullOrEmpty(sdkRequest.headers().get(CONNECTION))) {
+            crtHeaderList.add(new HttpHeader(CONNECTION, KEEP_ALIVE));
+        }
+
+        // Set Content-Length if needed
+        Optional<Long> contentLength = asyncRequest.requestContentPublisher().contentLength();
+        if (isNullOrEmpty(sdkRequest.headers().get(CONTENT_LENGTH)) && contentLength.isPresent()) {
+            crtHeaderList.add(new HttpHeader(CONTENT_LENGTH, Long.toString(contentLength.get())));
+        }
+
+        // Add the rest of the Headers
+        for (Map.Entry<String, List<String>> headerList: sdkRequest.headers().entrySet()) {
+            for (String val: headerList.getValue()) {
+                HttpHeader h = new HttpHeader(headerList.getKey(), val);
+                crtHeaderList.add(h);
+            }
+        }
+
+        return crtHeaderList;
+    }
+
+    private HttpHeader[] asArray(List<HttpHeader> crtHeaderList) {
+        return crtHeaderList.toArray(new HttpHeader[crtHeaderList.size()]);
+    }
+
+    private HttpRequest toCrtRequest(URI uri, AsyncExecuteRequest asyncRequest, AwsCrtAsyncHttpStreamAdapter crtToSdkAdapter) {
+        SdkHttpRequest sdkRequest = asyncRequest.request();
+        Validate.notNull(uri, "URI must not be null");
+        Validate.notNull(sdkRequest, "SdkHttpRequest must not be null");
+
+        String method = sdkRequest.method().name();
+        String encodedPath = sdkRequest.encodedPath();
+        String encodedQueryString = SdkHttpUtils.encodeAndFlattenQueryParameters(sdkRequest.rawQueryParameters())
+                .map(value -> "?" + value)
+                .orElse("");
+
+        HttpHeader[] crtHeaderArray = asArray(createHttpHeaderList(uri, asyncRequest));
+
+        return new HttpRequest(method, encodedPath + encodedQueryString, crtHeaderArray, crtToSdkAdapter);
+    }
+
+    @Override
+    public CompletableFuture<Void> execute(AsyncExecuteRequest asyncRequest) {
+        if (isClosed.get()) {
+            throw new IllegalStateException("Client is closed. No more requests can be made with this client.");
+        }
```

Comment:
Submitting a request as close is occurring (or after) feels like a serious usage error.  How far should we go to handle that for the user?

Reply:
Yes, I agree as well that it would be a usage error, but since the interface requires that implementations are threadsafe, we should account for this behavior.

Reply:
Hmm if close and getOrCreateConnectionPool need to be mutually exclusive then the simplest thing I can think of is drop the AtomicBoolean and use a RW Lock.  getOrCreate takes the reader lock and checks closed after, closed takes the write lock and only proceeds if the boolean is false.  The create path could be an upgrade from read -> write or a separate lock.